### PR TITLE
68 Polars based search

### LINKED ISSUE #68: Bug in `tests/test_search.py/test_search_columns_with_iterables`
# Describe the bug
In the 3rd parametrization of `tests/test_search/test_search_columns_with_iterables.py`, we have the following:
```python
def test_search_columns_with_iterables(query, require_all, expected):
    df = pd.DataFrame(
        {
            "A": ["cat0", "cat1", "cat1"],
            "B": [["a", "b"], ["a", "b", "c"], ["c", "d", "a"]],
            "C": [("cx", "cy"), ("cx", "cz"), ("cz", "cy")],
            "D": [{0, 1}, {0, 2}, {0, 1}],
            "E": ["xxx", "xxx", "yyy"],
        }
    )
    results = search(
        df=df,
        query=query,
        columns_with_iterables=["B", "C", "D"],
        name_column="A",
        require_all=require_all,
    ).to_dict(orient="records")
    assert results == expected
```
where 
```python 
query =  {'B': ['a', 'b'], 'D': [0]}
require_all =  True
```
This means we start with the following dataframe:
```python
      A          B         C       D    E
0  cat0     [a, b]  (cx, cy)  {0, 1}  xxx
1  cat1  [a, b, c]  (cx, cz)  {0, 2}  xxx
2  cat1  [c, d, a]  (cz, cy)  {0, 1}  yyy
```
and apply the search parameters above to it. In the test, we have 
```python
>>> expected
      A       B         C    D    E
0  cat0  [a, b]  (cx, cy)  {0}  xxx
1  cat1  [a, b]  (cx, cz)  {0}  xxx
2  cat1     [a]  (cz, cy)  {0}  yyy
```
ie. all rows match - however `df.loc[2, 'B']` returns a list filtered for elements contained in the query. However, we have `require_all=True`, and so as `df.loc[2, 'B'] != query['B']`, this row should not be in the returned dataframe.

I believe that the test should expect this returned dataframe
```python
      A       B         C    D    E
0  cat0  [a, b]  (cx, cy)  {0}  xxx
1  cat1  [a, b]  (cx, cz)  {0}  xxx
```
# Additional context
I came across this bug whilst trying to wrap my head around this package. The `search._search` function is pretty difficult to understand & so I rewrote it from scratch in [polars](https://docs.pola.rs/api/python/stable/reference/index.html) as an exercise in understanding how it functions. I've been unable to reconcile this test failure with the other tests in my reimplementation.

### PR DESCRIPTION
**Pull request by charles-turner-1**
Closes #68.

This PR is a bit of a monster & addresses the difficulty I had understanding the `search._search` function. It's hard to say for certain, but I think the complexity of this function lead to some erroneous/inconsistent tests.

I have:
- Refactored the search function to use [polars](https://docs.pola.rs/api/python/stable/reference/index.html) instead of pandas & numpy in order to (hopefully) simplify the logic.
- Removed & altered some tests which appear to be inconsistent with other tests & documentation.
- Added type hints everywhere, fixed inconsistent typing (mypy reports no errors).

The change from performing catalog searches using polars over pandas should improve performance pretty substantially. I haven't tested this explicitly yet (the tests only run on small catalogues), but polars is in general *much* more performant, in particular when using lazy evaluation as I've done here.

NB: Although catalog search has been changed to use polars in this PR, the user will never see a polars dataframe, only a pandas dataframe: conversion back and forth happens internally within the search functionality.

### CODE REVIEW

File: tests/test_core.py [hunk 1]
```diff
@@ -258,14 +258,12 @@ def test_catalog_keys(catalog_path):
     [
         ({"realm": "ocean"}, False, 1),
         ({"realm": ["atmos", "ocnBgchem"]}, False, 3),
-        ({"realm": ["atmos", "ocnBgchem"]}, True, 1),
```
Comment (charles-turner-1) [inline tests/test_core.py:261]: Realm is not defined to be a column with iterables here, and so the `require_all=True` argument doesn't really make sense.

Additionally, in the test dataset provided, only one realm is provided per dataset. If we were to interpret the `require_all=True` as meaning that we needed to match both 'atmos' and 'ocnBgchem', we wouldn't expect any results

File: tests/test_core.py [hunk 2]
```diff
@@ -258,14 +258,12 @@ def test_catalog_keys(catalog_path):
...
         ({"realm": "atmos"}, False, 3),
         ({"realm": "atmos", "variable": "tas"}, False, 1),
         ({"realm": "atmos", "variable": ["tas"]}, False, 1),
         ({"variable": ["NO2", "tas", "fgco2"]}, False, 3),
         ({"variable": ["NO2", "tas", "fgco2"]}, True, 0),
         ({"name": ["cesm", "cmip5"]}, False, 2),
-        ({"name": ["cesm", "cmip5"]}, True, 0),
```
Comment (charles-turner-1) [inline tests/test_core.py:268]: As above comment [regarding realm](https://github.com/ACCESS-NRI/intake-dataframe-catalog/pull/69/files#r1830384959)

File: tests/test_search.py [hunk 3]
```diff
@@ -176,7 +177,7 @@ def test_is_pattern(value, expected):
             ],
         ),
         (
-            {"A": [re.compile("^a.*a$", flags=re.IGNORECASE)]},
+            {"A": ["(?i)^a.*a$"]},
```
Comment (charles-turner-1) [inline tests/test_search.py:180]: Polars string matching has first class support for regexes. Setting case insensitive search in polars is as straightforward as adding a `(?i)` group.

There doesn't seem to be any documentation indicating that we expose compiled regex's to the user as a way to interact with the catalog, so I think it's safe to drop the use of `re.compile` here.

File: tests/test_search.py [hunk 4]
```diff
@@ -333,7 +333,7 @@ def test_search(query, expected):
                 },
                 {
                     "A": "cat1",
-                    "B": ["a", "c"],
+                    "B": ["c", "a"],
```
Comment (charles-turner-1) [inline tests/test_search.py:336]: When searching over list columns in polars, the ordering of the initial list column is preserved. 

Whether we would prefer to preserve the initial list order or sort in order of the query as the previous implementation did is an open question - I can see arguments for both.

I've altered the test as I don't think it's important & it reduces the amount of data processing necessary.

File: tests/test_search.py [hunk 5]
```diff
@@ -353,7 +353,7 @@ def test_search(query, expected):
...
```
Comment (charles-turner-1) [inline tests/test_search.py:356]: As above regarding [list ordering](https://github.com/ACCESS-NRI/intake-dataframe-catalog/pull/69/files#r1830388735)

File: tests/test_search.py [hunk 6]
```diff
@@ -258,7 +259,6 @@ def test_search(query, expected):
             [
                 {"A": "cat0", "B": ["a", "b"], "C": ("cx", "cy"), "D": {0}, "E": "xxx"},
                 {"A": "cat1", "B": ["a", "b"], "C": ("cx", "cz"), "D": {0}, "E": "xxx"},
-                {"A": "cat1", "B": ["a"], "C": ("cz", "cy"), "D": {0}, "E": "yyy"},
```
Comment (charles-turner-1) [inline tests/test_search.py:261]: I believe this test was erroneous: see #68

**Comment by charles-turner-1** [comment]
I've run a few performance tests - it looks like the polars based implementation is similarly performant on simple(ish) searches & quite a bit faster on more complicated searches: see below

```python
import intake
cat = intake.cat.access_nri

%%timeit
cat.search(variable='(?i).*temp.*')

# Using polars based implementation
# 30.4 ms ± 613 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)

# Using pandas based implementation
# 37 ms ± 934 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
cat.search(description='(?i).*access-om2.*|.*jra.*')

# Using polars based implementation
# 87.3 ms ± 2.61 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

# Using pandas based implementation
# 77 ms ± 1.83 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
cat.search(description='(?i).*access-om2.*|.*jra.*', variable='(?i).*global.*')

# Using polars based implementation
# 19.6 ms ± 163 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

# Using pandas based implementation
# 108 ms ± 1.6 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
cat.search(description='(?i).*access-om2.*|.*jra.*', variable='(?i).*global.*', realm='(?i)seaice')

# Using polars based implementation
# 12.3 ms ± 199 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

# Using pandas based implementation
# 112 ms ± 2.19 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

``

**Comment by charles-turner-1** [comment]
Since it's looking more probable we're going to use polars in intake-esm, I think it's probably time to merge this?

File: .conda/meta.yaml [hunk 7]
```diff
@@ -22,6 +22,7 @@ requirements:
     - python >=3.10
     - intake >=0.7.0
     - pandas
+    - polars >=1.12
```
Comment (marc-white) [inline .conda/meta.yaml:25]: Should test suite dependencies (like `intake_esm`) be added here?

(Incidentally, I'm getting the following test failure on both this branch and `main`, so I'm guessing I've got something that's not set up correctly...)

```
FAILED tests/test_core.py::test_read_source - intake_esm.source.ESMDataSourceError: Failed to load dataset with key='ocn.CTRL.pop.h'
```

Comment (charles-turner-1) [reply .conda/meta.yaml:25]: Re. Dependencies, I don't think so, because we don't want the distribution to include test only dependencies.

I've got no real idea what might be making that test fail - runs fine on my machine & CI, so I suspect it might be the wrong environment or similar activated on your machine?

File: tests/test_search.py (hunk 4 above)
Comment (dougiesquire) [reply tests/test_search.py:337]: Agreed

**Review by dougiesquire — COMMENTED** [review]
LGTM @charles-turner-1. I'll admit I haven't look deeply at the refactor, but my recollection is that the tests, though messy, were capturing the weird edge cases. They're still passing, so I'm happy.

One thought. It might be worth adding `polars>=1.12` to the environments in `ci/environment-3.*.yml` so that polars is installed via conda in the CI, rather than being installed by pip when `intake-dataframe-catalog` is installed.

File: .conda/meta.yaml (hunk 7 above)
Comment (dougiesquire) [reply .conda/meta.yaml:25]: Just noting that tests all ran fine locally for me too

**Comment by charles-turner-1** [comment]
> LGTM @charles-turner-1. I'll admit I haven't look deeply at the refactor, but my recollection is that the tests, though messy, were capturing the weird edge cases. They're still passing, so I'm happy.
> 
> One thought. It might be worth adding `polars>=1.12` to the environments in `ci/environment-3.*.yml` so that polars is installed via conda in the CI, rather than being installed by pip when `intake-dataframe-catalog` is installed.

Yeah, I think I had some confusion with the keys or something like that. I haven't included anything to make the tests easier to follow in this PR, it'd probably be worthwhile to do so at some point though.

**Review by dougiesquire — APPROVED** [review]
